[iOS] Fix crash in ProcessThrottler::sendPrepareToSuspendIPC

salinas-miguel opened the pull request:
#### 9bdfab144ce088159138cc60fe84bd20f293576a
<pre>
[iOS] Fix crash in ProcessThrottler::sendPrepareToSuspendIPC
<a href="https://bugs.webkit.org/show_bug.cgi?id=249220">https://bugs.webkit.org/show_bug.cgi?id=249220</a>
rdar://103276264

Reviewed by Chris Dumez.

After 257628@main we call `ProcessThrottler::updateThrottleStateIfNeeded`
every time we call `ProcessThrottler::setShouldTakeSuspendedAssertion`.
Instead, we should only take or drop the Suspended assertion when actually
needed in `ProcessThrottler::setShouldTakeSuspendedAssertion`.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setShouldTakeSuspendedAssertion):

Canonical link: <a href="https://commits.webkit.org/257850@main">https://commits.webkit.org/257850@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83e568a3a6baa44570a7cd14990eea2515cd1a6a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100131 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9300 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33205 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109465 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10183 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Big-Sur-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92562 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107355 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105900 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7696 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90979 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34398 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89602 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22374 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3068 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23889 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3040 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Big-Sur-Debug-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9170 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43374 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5389 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4878 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -211,7 +211,7 @@ void ProcessThrottler::updateThrottleStateIfNeeded()
         // and clean up - move it to the background and send it a message to notify. Schedule a timeout so it can't stay running
         // in the background for too long.
-        if (m_state != ProcessThrottleState::Suspended) {
+        if (m_state != ProcessThrottleState::Suspended && !m_pendingRequestToSuspendID) {
```
This change doesn't look right. We need to figure out what you changed recently to cause this crash and fix the underlying cause.

Your proposed fix would cause us to potentially release the process assertion while the PrepareToSuspend IPC handshake is still going on, AFAICT.

cdumez replied:
If we don't understand what caused this to start happening, I'd fall back to the following fix that should be safer:
```
if (m_state != ProcessThrottleState::Suspended) {
    if (!m_pendingRequestToSuspendID) {
        m_prepareToSuspendTimeoutTimer.startOneShot(processSuspensionTimeout);
        sendPrepareToSuspendIPC(IsSuspensionImminent::No);
    }
    return;
}
```

So we avoid sending the IPC (since we already sent one), but don't call `updateThrottleStateNow()` too early. Doing so while the IPC handshake is going on would cause us to release our process assertion and suspend I believe, before the WebProcess told us it was ready to suspend.

salinas-miguel replied:
The change that caused this is the newly added call to `updateThrottleStateIfNeeded()` in `setShouldTakeSuspendedAssertion()`. I think the right solution would actually be to remove the new call to `updateThrottleStateIfNeeded()` and instead just drop/take the suspended assertion if needed

cdumez reviewed (approved):
Looks like the commit message needs updating.

salinas-miguel committed:
[iOS] Fix crash in ProcessThrottler::sendPrepareToSuspendIPC
https://bugs.webkit.org/show_bug.cgi?id=249220
rdar://103276264

Reviewed by Chris Dumez.

After 257628@main we call `ProcessThrottler::updateThrottleStateIfNeeded`
every time we call `ProcessThrottler::setShouldTakeSuspendedAssertion`.
Instead, we should only take or drop the Suspended assertion when actually
needed in `ProcessThrottler::setShouldTakeSuspendedAssertion`.

* Source/WebKit/UIProcess/ProcessThrottler.cpp:
(WebKit::ProcessThrottler::setShouldTakeSuspendedAssertion):

Canonical link: https://commits.webkit.org/257850@main
